### PR TITLE
New JMESPATH extractor

### PR DESCRIPTION
**Pull request by ubikloadpack**
## Description

UbikLoadPack team is happy to contribute this new component that allows using JMESPATH (http://jmespath.org/) as a new technology to extract data from JSON.

Regards
Qian Qian
Philippe 

## Motivation and Context

JSON-PATH is available today but has many limits:
- As seen here https://bz.apache.org/bugzilla/show_bug.cgi?id=63623, the Java library does not offer what JSON-PATH in other implementation provides which confuses users
- JMESPATH has a specification, implementations in different languages and a test suite
- It is more powerful

## How Has This Been Tested?

JUnit Test case
Manual Testing

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist:

- [X] My code follows the [code style][style-guide] of this project.
- [X] I have updated the documentation accordingly.
- [X] Add elements in saveservice.properties
- [X] Remove double quotes around string values

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

### CODE REVIEW

File: src/bom/build.gradle.kts [hunk 1]
```diff
@@ -77,6 +77,8 @@ dependencies {
         apiv("commons-lang:commons-lang")
         apiv("commons-net:commons-net")
         apiv("dnsjava:dnsjava")
+	apiv("io.burt:jmespath-core")
```
Comment (vlsi) [inline src/bom/build.gradle.kts]: Please refrain from using tabs.
We should probably setup a checkstyle verification

**Comment by vlsi** [comment]
Please fix Travis failures as well

**Comment by ubikloadpack** [comment]
> Please fix Travis failures as well

Last failure has no relation with PR, fixed by pmouawad commit.

Other ones fixed this morning

**Comment by vlsi** [comment]
Let me clarify.
Please fix violations in JMESPathExtractor  and JMESPathExtractorGui: https://travis-ci.org/apache/jmeter/jobs/581079105#L390-L400

File: src/components/src/main/java/org/apache/jmeter/extractor/json/jmespath/JMESPathExtractor.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+            if (log.isDebugEnabled()) {
+                log.debug("Error processing JSON content in {}, message: {}", getName(), e.getLocalizedMessage(), e);
+            } else {
+                log.debug("Error processing JSON content in {}, message: {}", getName(), e.getLocalizedMessage());
+            }
+            vars.put(refName, defaultValue);
+        }
+    }
+
+    public List<String> splitJson(JsonNode jsonNode) throws JsonParseException, JsonMappingException, IOException {
```
Comment (FSchumacher) [inline src/components/src/main/java/org/apache/jmeter/extractor/json/jmespath/JMESPathExtractor.java]: The exceptions are subclasses of `IOException`. Do we  really need to specify them here?

File: src/components/src/main/java/org/apache/jmeter/extractor/json/jmespath/JMESPathExtractor.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        List<String> splittedJsonElements = new ArrayList<>();
+        ObjectMapper mapper = new ObjectMapper();
+        if (jsonNode.isArray()) {
+            ArrayNode arrayNode = (ArrayNode) jsonNode;
+            for (int i = 0; i < arrayNode.size(); i++) {
```
Comment (FSchumacher) [inline src/components/src/main/java/org/apache/jmeter/extractor/json/jmespath/JMESPathExtractor.java]: Why not use a for loop like
```
for (JsonNode element : ((ArrayNode) jsonNode)) {
    splittedJsonElements.add(mapper.writeValueAsString(element));
}
```

File: src/components/src/test/java/org/apache/jmeter/extractor/json/jmespath/TestJMESPathExtractor.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    public void testJMESPathExtractorAllElementsOneMatch() {
+        // test1
+        JMeterContext context = JMeterContextService.getContext();
+        JMESPathExtractor processor = setupProcessor(context, "-1");
+        JMeterVariables vars = new JMeterVariables();
+        processor.setJmesPathExpression("[*]");
+        context.setVariables(vars);
+        vars.put("contentvar", "[\"one\"]");
+        processor.process();
+        assertThat(vars.get(REFERENCE_NAME), CoreMatchers.is(CoreMatchers.nullValue()));
```
Comment (FSchumacher) [inline src/components/src/test/java/org/apache/jmeter/extractor/json/jmespath/TestJMESPathExtractor.java]: Maybe this assert sequence could be extracted into a helper assert method like `assertValues(vars, values...)`?

File: src/components/src/test/java/org/apache/jmeter/extractor/json/jmespath/TestJMESPathExtractor.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        context.setVariables(vars);
+        processor.setJmesPathExpression("a.b.c.f");
+        vars.put("contentvar", "{\"a\": {\"b\": {\"c\": {\"d\": \"value\"}}}}");
+        processor.process();
+        assertThat(vars.get(REFERENCE_NAME), CoreMatchers.is(DEFAULT_VALUE));
+        assertThat(vars.get(REFERENCE_NAME_MATCH_NUMBER), CoreMatchers.is("0"));
+    }
+
+    @Test
+    public void testJMESPathExtractorAllElementsOneMatch() {
```
Comment (FSchumacher) [inline src/components/src/test/java/org/apache/jmeter/extractor/json/jmespath/TestJMESPathExtractor.java]: This seems to be a quite complex test, would it be better to split the tested components into simpler tests?

File: src/components/src/test/java/org/apache/jmeter/extractor/json/jmespath/TestJMESPathExtractor.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+                "{\r\n" + "  \"people\": [\r\n" + "    {\"first\": \"James\", \"last\": \"d\"},\r\n"
+                        + "    {\"first\": \"Jacob\", \"last\": \"e\"},\r\n"
+                        + "    {\"first\": \"Jayden\", \"last\": \"f\"},\r\n" + "    {\"missing\": \"different\"}\r\n"
+                        + "  ],\r\n" + "  \"foo\": {\"bar\": \"baz\"}\r\n" + "}");
+        processor.process();
+        assertThat(vars.get(REFERENCE_NAME + "_1"), CoreMatchers.is("{\"first\":\"Jayden\",\"last\":\"f\"}"));
+    }
+
+    @Test
+    public void testJMESPathExtractorAllElementsMultipleMatches() {
```
Comment (FSchumacher) [inline src/components/src/test/java/org/apache/jmeter/extractor/json/jmespath/TestJMESPathExtractor.java]: Like the test before, would it not be better to split it into simpler tests?

**Comment by vlsi** [comment]
I see @FSchumacher has added relevant license files to the JMeter repository, however it would be nice if we could ask upstream project to include LICENSE file as `META-INF/LICENSE`.

See samples here: https://github.com/apache/jmeter/pull/469#issuecomment-502834212

@pmouawad would you please ask jmespath for shipping relevant license inside jmespath releases so we could get always up to date version of it?

PS. I've added dependency license verification to Travis as well.

**Pull request merged**